### DISCUSSION TITLE
Remote API detector

### DISCUSSION THREAD

**Discussion by RichardPar** [blakeblackshear/frigate#5811] [Ideas]
As frigate uses Coral - and seems to lean towards that system; Could a remote API be defined whereby a JSON message be sent to a remote AI system for image processing using IP sockets. (Such as DOODS or equivalent) 

This allows for other AI systems to be used and experimented with - opens flexability for more experimentation. 

Would this be possible?

Richard

**Comment by NickM-27** [comment]
No, the latency is too high

https://github.com/blakeblackshear/frigate/issues/5766

**Reply by RichardPar** [reply]
I disagree in all senses!  

Why is latency an issue? What is being achieved aiming for zero latency? .. if it takes 500mSec - what is the issue?  If Latency was a real issue, then MQTT would never have been used either.

I have a Jetson doing it in under 10mSeconds over IP from another system.. and Jetson is horrible! :D 

If the real answer is 'I dont want to' - then I would go quietly in to the night, but using 'latency' is a weak technical argument.

**Reply by NickM-27** [reply]
> I disagree in all senses!
> 
> Why is latency an issue? What is being achieved aiming for zero latency? .. if it takes 500mSec - what is the issue? If Latency was a real issue, then MQTT would never have been used either.

Frigate is built with object detection at its core and it is very focused on real time detection. If an inference took 500 ms then you could only run detection on two frames per second on a single camera to remain real time. This is one reason why using CPU detectors works poorly, if your inference time is even 200ms with detect running at 5 fps then there is a high likelihood that some frames will be skipped during periods of high activity. 

https://github.com/blakeblackshear/frigate/issues/3449 exists as a proof of concept and it sees detection times of 50 ms with a coral and that is with raw TCP.

It's also unclear why you mention MQTT as MQTT has nothing to do with Frigate's detection pipeline.

**Reply by RichardPar** [reply]
> > I disagree in all senses!
> > Why is latency an issue? What is being achieved aiming for zero latency? .. if it takes 500mSec - what is the issue? If Latency was a real issue, then MQTT would never have been used either.
> 
> Frigate is built with object detection at its core and it is very focused on real time detection. If an inference took 500 ms then you could only run detection on two frames per second on a single camera to remain real time. This is one reason why using CPU detectors works poorly, if your inference time is even 200ms with detect running at 5 fps then there is a high likelihood that some frames will be skipped during periods of high activity.
> 
> #3449 exists as a proof of concept and it sees detection times of 50 ms with a coral and that is with raw TCP.
> 
> It's also unclear why you mention MQTT as MQTT has nothing to do with Frigate's detection pipeline.

MQTT is mentioned as that is what gets the events from Frigate -  Frigate is a sports-car that is driving on cobble-stone roads (MQTT being the road) 

You misunderstand how a good IP based implementation works... You can run multiple threads/proceses, you dont have to be sequential. Frigate can connect over 10 sockets and get 10 replies back in the same 500mSecond timeframe.

Sure, I agree with realtime - but to what extent ? .. Is 500mSec .. or ever 1 second going to have any impact on the end experience.. what usecase scanario is REALTIME a requirement?

**Reply by NickM-27** [reply]
> MQTT is mentioned as that is what gets the events from Frigate - Frigate is a sports-car that is driving on cobble-stone roads (MQTT being the road)

Sure, but that is completely unrelated to Frigate's detection pipeline which means it is irrelevant in the conversation of detection itself.

> You misunderstand how a good IP based implementation works... You can run multiple threads/proceses, you dont have to be sequential. Frigate can connect over 10 sockets and get 10 replies back in the same 500mSecond timeframe.

Detectors like the coral are purely sequential, they can only process one image at a time. So while the communication can be multithreaded, there is still a sequential bottleneck on the detector itself. You would also need to get the results of the detections back in the order they were sent or object tracking would not work at all. There is also the concern of inconsistent latency. As the remote detector gets a larger and larger queue from multiple cameras, the detection times will be very inconsistent leading to issues as well. 

It would likely be a large amount of effort that would require a lot of edge case handling all to get to a point which has inference times worse than what can be achieved with OpenVINO using an integrated GPU.
 
> Sure, I agree with realtime - but to what extent ? .. Is 500mSec .. or ever 1 second going to have any impact on the end experience.. what usecase scanario is REALTIME a requirement?

The realtime is not a decision just for user experience, it is inherent to the way frigate's processing works.

**Reply by RichardPar** [reply]
> > MQTT is mentioned as that is what gets the events from Frigate - Frigate is a sports-car that is driving on cobble-stone roads (MQTT being the road)
> 
> Sure, but that is completely unrelated to Frigate's detection pipeline which means it is irrelevant in the conversation of detection itself.
> 
> > You misunderstand how a good IP based implementation works... You can run multiple threads/proceses, you dont have to be sequential. Frigate can connect over 10 sockets and get 10 replies back in the same 500mSecond timeframe.
> 
> Detectors like the coral are purely sequential, they can only process one image at a time. So while the communication can be multithreaded, there is still a sequential bottleneck on the detector itself. You would also need to get the results of the detections back in the order they were sent or object tracking would not work at all. There is also the concern of inconsistent latency. As the remote detector gets a larger and larger queue from multiple cameras, the detection times will be very inconsistent leading to issues as well.
> 

That is true, Coral is squeintial - but IP sockets are not - you can queue up frames in the background, then marshall them - Coral is the quickest part of the equasion. 

> It would likely be a large amount of effort that would require a lot of edge case handling all to get to a point which has inference times worse than what can be achieved with OpenVINO using an integrated GPU.
> 
> > Sure, I agree with realtime - but to what extent ? .. Is 500mSec .. or ever 1 second going to have any impact on the end experience.. what usecase scanario is REALTIME a requirement?
> 
> The realtime is not a decision just for user experience, it is inherent to the way frigate's processing works.

Thanks for the conversation - I still disagree on many points you have,  but then I didnt write Frigate. 

My main concern/idea  is that I would like to add more detectors, but the Author states he doesnt want to support anything other than Coral. (and I am not a Python developer!)  I would like to break that dependency and atleast have a simple API whereby anyone is empowered to add their own plugin in a simple manner.. without begging the Frigate masters.

**Reply by NickM-27** [reply]
> Thanks for the conversation - I still disagree on many points you have,  but then I didnt write Frigate. 
> 
> My main concern/idea  is that I would like to add more detectors, but the Author states he doesnt want to support anything other than Coral.

That's entirely untrue, Frigate 0.12 supports coral, intel NCS2 via openvino, intel CPU via openvino, intel GPU via openvino, and nvidia GPU via tensorrt. 

https://deploy-preview-4055--frigate-docs.netlify.app/configuration/detectors

There's also PRs up to add support for ARM devices with Mali GPUs and support for RKNN accelerator that is built into orange pi devices. 

https://github.com/blakeblackshear/frigate/pull/5733

**Reply by RichardPar** [reply]
> That's entirely untrue, Frigate 0.12 supports coral, intel ncs2 via openvino, intel cpu via openvino, intel GPU via openvino, and nvidia GPU via tensorrt.
> 
> https://deploy-preview-4055--frigate-docs.netlify.app/configuration/detectors
> 
> There's also PRs up to add support for ARM devices with Mali GPUs and support for RKNN accelerator that is built into orange pi devices.

Thanks! will update my memory - When I asked the question a few years ago, the reponse was different :D 

https://github.com/blakeblackshear/frigate/discussions/1109
 :D

I have got 3 new development boards that have new NPU's (which are not on the list) .. and I would like to contribute
